LibraBFT/Impl/Consensus/BlockStorage/SyncManager.agda

```diff
@@ -4,20 +4,60 @@
    Licensed under the Universal Permissive License v 1.0 as shown at https://opensource.oracle.com/licenses/upl
 -}
 
-open import LibraBFT.Base.ByteString
-open import LibraBFT.Base.PKCS
-open import LibraBFT.Base.Types
-open import LibraBFT.Hash
-open import LibraBFT.Impl.Consensus.ConsensusTypes.Vote as Vote
-open import LibraBFT.ImplShared.Base.Types
+open import LibraBFT.Impl.Consensus.BlockStorage.BlockStore   as BlockStore
+import      LibraBFT.Impl.Consensus.BlockStorage.BlockTree    as BlockTree
+open import LibraBFT.Impl.Consensus.ConsensusTypes.Vote       as Vote
+open import LibraBFT.Impl.OBM.Logging.Logging
 open import LibraBFT.ImplShared.Consensus.Types
-open import LibraBFT.ImplShared.Util.Crypto
 open import LibraBFT.ImplShared.Util.Util
-open import LibraBFT.Abstract.Types.EpochConfig UID NodeId
 open import LibraBFT.Prelude
 open import Optics.All
 
 module LibraBFT.Impl.Consensus.BlockStorage.SyncManager where
 
+data NeedFetchResult : Set where
+  QCRoundBeforeRoot QCAlreadyExist QCBlockExist NeedFetch : NeedFetchResult
+
 postulate
-  insertQuorumCertM : QuorumCert → BlockRetriever → LBFT (Either ErrLog Unit)
+
+  liNextEpochState : Lens LedgerInfo (Maybe EpochState)
+  qcEndsEpoch      : Lens QuorumCert Bool
+
+  fetchQuorumCertM
+    : QuorumCert → BlockRetriever
+    → LBFT (Either ErrLog Unit)
+
+  needFetchForQuorumCert
+    : QuorumCert → BlockStore
+    → Either ErrLog NeedFetchResult
+
+------------------------------------------------------------------------------
+
+insertQuorumCertM
+  : QuorumCert → BlockRetriever
+  → LBFT (Either ErrLog Unit)
+insertQuorumCertM qc retriever = do
+  bs ← use lBlockStore
+  _ ← case needFetchForQuorumCert qc bs of \where
+    (Left e) →
+      bail e
+    (Right NeedFetch) →
+      fetchQuorumCertM qc retriever
+      ∙^∙ withErrCtxt
+    (Right QCBlockExist) →
+      BlockStore.insertSingleQuorumCertM qc ∙^∙ withErrCtxt ∙?∙ λ _ → do
+      ok unit
+    (Right _) →
+      ok unit
+  maybeS (bs ^∙ bsRoot) (bail fakeErr) $ λ bsr →
+    ifM (bsr ^∙ ebRound) <?ℕ (qc ^∙ qcCommitInfo ∙ biRound)
+      then (do
+        let finalityProof = qc ^∙ qcLedgerInfo
+        BlockStore.commitM finalityProof {- (.?. (λ xx →
+          if qc ^∙ qcEndsEpoch
+            then ok unit -- TODO-1 EPOCH CHANGE
```

Comment:
It  makes sense not to model epoch change yet, but I'd prefer to have the rest of this code, and just leave `ok unit  -- TODO-1 EPOCH CHANGE`. It means that the proofs will break less when we do model epoch change.

Reply:
DONE